File: engine/src/main/java/com/shatteredpixel/engine/geom/Rect.java
package com.shatteredpixel.engine.geom;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Immutable axis-aligned rectangle on an integer grid.
 * Represented by (left, top, right, bottom) where right and bottom are exclusive.
 *
 * Migrated from: com.watabou.utils.Rect
 * Changes:
 * - Made immutable (final fields, returns new instances)
 * - Deterministic center() (no Random usage)
 * - Added proper contains/intersects checks
 * - Added area() method
 * - GWT-safe (Java 11)
 */
public final class Rect {

    public final int left;
    public final int top;
    public final int right;
    public final int bottom;

    // =========================================================================
    // Constructors
    // =========================================================================

    public Rect(int left, int top, int right, int bottom) {
        this.left = left;
        this.top = top;
        this.right = right;
        this.bottom = bottom;
    }

    /**
     * Create from position and size.
     */
    public static Rect fromSize(int x, int y, int width, int height) {
        return new Rect(x, y, x + width, y + height);
    }

    /**
     * Create from two points (min and max corners).
     */
    public static Rect fromPoints(Point p1, Point p2) {
        int minX = Math.min(p1.x, p2.x);
        int minY = Math.min(p1.y, p2.y);
        int maxX = Math.max(p1.x, p2.x);
        int maxY = Math.max(p1.y, p2.y);
        return new Rect(minX, minY, maxX + 1, maxY + 1);
    }

    /**
     * Empty rectangle at origin.
     */
    public static Rect empty() {
        return new Rect(0, 0, 0, 0);
    }

    // =========================================================================
    // Properties
    // =========================================================================

    public int width() {
        return right - left;
    }

    public int height() {
        return bottom - top;
    }

    public int area() {
        return width() * height();
    }

    public boolean isEmpty() {
        return right <= left || bottom <= top;
    }

    // =========================================================================
    // Queries
    // =========================================================================

    /**
     * Check if this rectangle contains a point.
     */
    public boolean contains(Point p) {
        return contains(p.x, p.y);
    }

    /**
     * Check if this rectangle contains coordinates.
     */
    public boolean contains(int x, int y) {
        return x >= left && x < right && y >= top && y < bottom;
    }

    /**
     * Check if this rectangle fully contains another rectangle.
     */
    public boolean contains(Rect other) {
        return left <= other.left && right >= other.right &&
               top <= other.top && bottom >= other.bottom;
    }

    /**
     * Check if this rectangle intersects another rectangle.
     */
    public boolean intersects(Rect other) {
        return left < other.right && right > other.left &&
               top < other.bottom && bottom > other.top;
    }

    // =========================================================================
    // Operations (return new instances - immutable)
    // =========================================================================

    /**
     * Get the center point of this rectangle.
     * For even dimensions, returns the lower-left center cell.
     */
    public Point center() {
        return new Point((left + right) / 2, (top + bottom) / 2);
    }

    /**
     * Shrink this rectangle by d cells on all sides.
     */
    public Rect shrink(int d) {
        return new Rect(left + d, top + d, right - d, bottom - d);
    }

    /**
     * Shrink by 1 cell on all sides.
     */
    public Rect shrink() {
        return shrink(1);
    }

    /**
     * Expand this rectangle by d cells on all sides.
     */
    public Rect expand(int d) {
        return new Rect(left - d, top - d, right + d, bottom + d);
    }

    /**
     * Expand by 1 cell on all sides.
     */
    public Rect expand() {
        return expand(1);
    }

    /**
     * Move this rectangle to a new position (preserves size).
     */
    public Rect moveTo(int x, int y) {
        int w = width();
        int h = height();
        return new Rect(x, y, x + w, y + h);
    }

    /**
     * Move this rectangle to a new position.
     */
    public Rect moveTo(Point p) {
        return moveTo(p.x, p.y);
    }

    /**
     * Shift this rectangle by an offset.
     */
    public Rect shift(int dx, int dy) {
        return new Rect(left + dx, top + dy, right + dx, bottom + dy);
    }

    /**
     * Shift this rectangle by a point offset.
     */
    public Rect shift(Point offset) {
        return shift(offset.x, offset.y);
    }

    /**
     * Resize this rectangle (keeps top-left corner fixed).
     */
    public Rect resize(int newWidth, int newHeight) {
        return new Rect(left, top, left + newWidth, top + newHeight);
    }

    /**
     * Scale all coordinates by a factor.
     */
    public Rect scale(int factor) {
        return new Rect(left * factor, top * factor, right * factor, bottom * factor);
    }

    /**
     * Intersection of this rectangle with another.
     * Returns empty rect if no intersection.
     */
    public Rect intersection(Rect other) {
        int l = Math.max(left, other.left);
        int t = Math.max(top, other.top);
        int r = Math.min(right, other.right);
        int b = Math.min(bottom, other.bottom);
        return new Rect(l, t, r, b);
    }

    /**
     * Union of this rectangle with another (smallest rect containing both).
     */
    public Rect union(Rect other) {
        if (isEmpty()) return other;
        if (other.isEmpty()) return this;

        int l = Math.min(left, other.left);
        int t = Math.min(top, other.top);
        int r = Math.max(right, other.right);
        int b = Math.max(bottom, other.bottom);
        return new Rect(l, t, r, b);
    }

    /**
     * Union with a single point (expand to include it).
     */
    public Rect union(Point p) {
        if (isEmpty()) {
            return new Rect(p.x, p.y, p.x + 1, p.y + 1);
        }

        int l = Math.min(left, p.x);
        int t = Math.min(top, p.y);
        int r = Math.max(right, p.x + 1);
        int b = Math.max(bottom, p.y + 1);
        return new Rect(l, t, r, b);
    }

    /**
     * Get all points inside this rectangle.
     */
    public List<Point> getAllPoints() {
        List<Point> points = new ArrayList<>();
        for (int y = top; y < bottom; y++) {
            for (int x = left; x < right; x++) {
                points.add(new Point(x, y));
            }
        }
        return points;
    }

    /**
     * Get all points on the border of this rectangle.
     */
    public List<Point> getBorderPoints() {
        List<Point> points = new ArrayList<>();
        if (isEmpty()) return points;

        // Top and bottom edges
        for (int x = left; x < right; x++) {
            points.add(new Point(x, top));
            if (height() > 1) {
                points.add(new Point(x, bottom - 1));
            }
        }

        // Left and right edges (skip corners already added)
        for (int y = top + 1; y < bottom - 1; y++) {
            points.add(new Point(left, y));
            if (width() > 1) {
                points.add(new Point(right - 1, y));
            }
        }

        return points;
    }

    // =========================================================================
    // Object overrides
    // =========================================================================

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (!(obj instanceof Rect)) return false;
        Rect other = (Rect) obj;
        return left == other.left && top == other.top &&
               right == other.right && bottom == other.bottom;
    }

    @Override
    public int hashCode() {
        return Objects.hash(left, top, right, bottom);
    }

    @Override
    public String toString() {
        return "[(" + left + "," + top + ")-(" + right + "," + bottom + ") " +
               width() + "x" + height() + "]";
    }
}

File: pixel-dungeon-web/src/utils/Rect.js
/*
 * Pixel Dungeon Web Port
 * Ported from Shattered Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 * Copyright (C) 2014-2025 Evan Debenham
 *
 * 2D Rectangle
 * Source: SPD-classes/src/main/java/com/watabou/utils/Rect.java
 */

import { Point } from './Point.js';
import { Random } from './Random.js';

export class Rect {
    constructor(left = 0, top = 0, right = 0, bottom = 0) {
        if (typeof left === 'object' && left instanceof Rect) {
            // Copy constructor: Rect.java:37-39
            this.left = left.left;
            this.top = left.top;
            this.right = left.right;
            this.bottom = left.bottom;
        } else {
            // Normal constructor: Rect.java:41-46
            this.left = left;
            this.top = top;
            this.right = right;
            this.bottom = bottom;
        }
    }

    /**
     * Calculate width
     * Source: Rect.java:48-50
     */
    width() {
        return this.right - this.left;
    }

    /**
     * Calculate height
     * Source: Rect.java:52-54
     */
    height() {
        return this.bottom - this.top;
    }

    /**
     * Calculate area
     * Source: Rect.java:56-58
     */
    square() {
        return this.width() * this.height();
    }

    /**
     * Set rectangle bounds
     * Source: Rect.java:60-70
     */
    set(left, top, right, bottom) {
        if (typeof left === 'object' && left instanceof Rect) {
            this.left = left.left;
            this.top = left.top;
            this.right = left.right;
            this.bottom = left.bottom;
        } else {
            this.left = left;
            this.top = top;
            this.right = right;
            this.bottom = bottom;
        }
        return this;
    }

    /**
     * Set position (keeping size constant)
     * Source: Rect.java:72-74
     */
    setPos(x, y) {
        return this.set(x, y, x + (this.right - this.left), y + (this.bottom - this.top));
    }

    /**
     * Shift rectangle by offset
     * Source: Rect.java:76-78
     */
    shift(x, y) {
        return this.set(this.left + x, this.top + y, this.right + x, this.bottom + y);
    }

    /**
     * Resize rectangle
     * Source: Rect.java:80-82
     */
    resize(w, h) {
        return this.set(this.left, this.top, this.left + w, this.top + h);
    }

    /**
     * Check if rectangle is empty
     * Source: Rect.java:84-86
     */
    isEmpty() {
        return this.right <= this.left || this.bottom <= this.top;
    }

    /**
     * Set rectangle to empty
     * Source: Rect.java:88-91
     */
    setEmpty() {
        this.left = this.right = this.top = this.bottom = 0;
        return this;
    }

    /**
     * Calculate intersection with another rectangle
     * Source: Rect.java:93-100
     */
    intersect(other) {
        const result = new Rect();
        result.left = Math.max(this.left, other.left);
        result.right = Math.min(this.right, other.right);
        result.top = Math.max(this.top, other.top);
        result.bottom = Math.min(this.bottom, other.bottom);
        return result;
    }

    /**
     * Calculate union with another rectangle
     * Source: Rect.java:102-109
     */
    union(other) {
        if (typeof other === 'object' && other instanceof Rect) {
            const result = new Rect();
            result.left = Math.min(this.left, other.left);
            result.right = Math.max(this.right, other.right);
            result.top = Math.min(this.top, other.top);
            result.bottom = Math.max(this.bottom, other.bottom);
            return result;
        } else if (typeof other === 'object' && other instanceof Point) {
            // Union with Point: Rect.java:129-131
            return this.union_xy(other.x, other.y);
        } else {
            // Union with x, y coordinates: Rect.java:111-127
            const x = other;
            const y = arguments[1];
            return this.union_xy(x, y);
        }
    }

    /**
     * Union with x, y coordinates
     * Source: Rect.java:111-127
     */
    union_xy(x, y) {
        if (this.isEmpty()) {
            return this.set(x, y, x + 1, y + 1);
        } else {
            if (x < this.left) {
                this.left = x;
            } else if (x >= this.right) {
                this.right = x + 1;
            }
            if (y < this.top) {
                this.top = y;
            } else if (y >= this.bottom) {
                this.bottom = y + 1;
            }
            return this;
        }
    }

    /**
     * Check if point is inside rectangle
     * Source: Rect.java:133-135
     */
    inside(p) {
        return p.x >= this.left && p.x < this.right && p.y >= this.top && p.y < this.bottom;
    }

    /**
     * Get center point (with random offset for even dimensions)
     * Source: Rect.java:137-141
     *
     * ⚠️ CRITICAL: This uses Random.Int for even-dimension randomization
     * Line 139-140 in Rect.java adds random offset when dimensions are even
     */
    center() {
        return new Point(
            Math.floor((this.left + this.right) / 2) + (((this.right - this.left) % 2) === 0 ? Random.Int_max(2) : 0),
            Math.floor((this.top + this.bottom) / 2) + (((this.bottom - this.top) % 2) === 0 ? Random.Int_max(2) : 0)
        );
    }

    /**
     * Shrink rectangle by distance d
     * Source: Rect.java:143-149
     */
    shrink(d = 1) {
        return new Rect(this.left + d, this.top + d, this.right - d, this.bottom - d);
    }

    /**
     * Scale rectangle by factor d
     * Source: Rect.java:151-153
     */
    scale(d) {
        return new Rect(this.left * d, this.top * d, this.right * d, this.bottom * d);
    }

    /**
     * Get all points within rectangle
     * Source: Rect.java:155-161
     */
    getPoints() {
        const points = [];
        for (let i = this.left; i <= this.right; i++) {
            for (let j = this.top; j <= this.bottom; j++) {
                points.push(new Point(i, j));
            }
        }
        return points;
    }

    /**
     * String representation
     */
    toString() {
        return `Rect(${this.left}, ${this.top}, ${this.right}, ${this.bottom})`;
    }
}
